<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Identify differentially expressed genes counts data • sageseqr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Identify differentially expressed genes counts data">
<meta property="og:description" content="This workflow normalizes counts, performs model selection from known 
    covariates using multivariate regression, automates outlier analysis using a 
    customizable threshold and performs differential expression analysis for repeated
    measures.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">sageseqr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/customize-config.html">Write your own config file</a>
    </li>
    <li>
      <a href="articles/run-the-plan.html">Run the drake plan</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Sage-Bionetworks/sageseqr/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="installation" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1></div>
<p><code><a href="https://remotes.r-lib.org/reference/install_github.html">remotes::install_github("Sage-Bionetworks/sageseqr")</a></code></p>
</div>
<div id="rna-seq-normalization-workflow-in-r" class="section level1">
<h1 class="hasAnchor">
<a href="#rna-seq-normalization-workflow-in-r" class="anchor"></a>RNA-seq normalization workflow in R</h1>
<p>The <code>sageseqr</code> package integrates the <a href="https://github.com/ropensci/drake/"><code>drake</code> R package</a>, the <a href="https://cran.r-project.org/web/packages/config/vignettes/introduction.html"><code>config</code> package for R</a>, and <a href="https://www.synapse.org/">Synapse</a>. <code>drake</code> tracks dependency relationships in the workflow and only updates data when it has changed. A <code>config</code> file allows inputs and parameters to be explicitly defined in one location. Synapse is a data repository that allows sensitive data to be <a href="https://docs.synapse.org/articles/article_index.html#governance">stored and shared responsibly</a>.</p>
<p>The workflow takes RNA-seq gene counts and sample metadata as inputs, normalizes counts by conditional quantile normalization <a href="https://bioconductor.org/packages/release/bioc/html/cqn.html">(CQN)</a>, removes outliers based on a user-defined threshold, empirically selects meaningful covariates and returns differential expression analysis results. The data is also visualized in several ways to help you understand meaningful trends. The visualizations include a heatmap identifying highly correlated covariates, a sample-specific x and y marker gene check, boxplots visualizing the distribution of continuous variables and a principal component analysis (PCA) to visualize sample distribution.</p>
</div>
<div id="the-targets" class="section level1">
<h1 class="hasAnchor">
<a href="#the-targets" class="anchor"></a>The Targets</h1>
<p>The series of steps that make up the workflow are called targets. The target objects are stored in a cache and can either be read or loaded into your environment with the <code>drake</code> functions <code>readd</code> or <code>loadd</code>. Source code for each target can be visualized by setting <code>show_source = TRUE</code> with <code>loadd</code> and <code>readd</code>.</p>
<p>Importantly, running <code>clean</code> will remove the data stored as targets (<a href="https://books.ropensci.org/drake/walkthrough.html">but, the data is never completely gone!</a>). You may specific targets by name by passing them to the <code>clean</code> function.</p>
<p>The targets are called by the <code>sageseqr</code> <code><a href="reference/rnaseq_plan.html">rnaseq_plan()</a></code> function and are:</p>
<p>Raw data:</p>
<ul>
<li>
<code>import_metadata</code>- imports the raw metadata directly from synapse</li>
<li>
<code>import_counts</code> - imports the raw counts directly from synapse</li>
<li>
<code>biomart_results</code> - the complete list of genes with biomaRt annotations.</li>
</ul>
<p>Exploratory data visualizations:</p>
<ul>
<li>
<code>gene_coexpression</code> - the distribution of correlated gene counts.</li>
<li>
<code>boxplots</code> - the distribution of continuous variables.</li>
<li>
<code>sex_plot</code> - the distribution of samples by x and y marker genes.</li>
<li>
<code>correlation_plot</code> - the correlation of covariates.</li>
<li>
<code>significant_covariates_plot</code> - the correlation of covariates to gene expression.</li>
<li>
<code>outliers</code> - the clustering of samples by PCA.</li>
</ul>
<p>Transformed or normalized data:</p>
<ul>
<li>
<code>clean_md</code> - metadata with factor and numeric types.</li>
<li>
<code>filtered_counts</code> - counts matrix with low gene expression removed.</li>
<li>
<code>biotypes</code> - gene proportions summarized by biotype.</li>
<li>
<code>cqn_counts</code> - CQN normalized counts.</li>
<li>
<code>model</code> - Model selected by multivariate forward stepwise regression (evaluated by Bayesian Information Criteria (BIC)).</li>
</ul>
</div>
<div id="access-to-data" class="section level1">
<h1 class="hasAnchor">
<a href="#access-to-data" class="anchor"></a>Access to Data</h1>
<p>Anyone can create a <a href="https://docs.synapse.org/articles/getting_started.html">Synapse account</a> and access public data in a variety of disciplines: <a href="https://adknowledgeportal.synapse.org/">Alzheimer’s Disease Knowledge portal</a>, <a href="https://www.synapse.org/#!Synapse:syn2759792/wiki/69613">CommonMind Consoritum</a>.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/Sage-Bionetworks/sageseqr/">https://​github.com/​Sage-Bionetworks/​sageseqr/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Kelsey Montgomery <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0126-8944" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Jake Gockley <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/kelshmo/sageseqr/actions"><img src="https://github.com/kelshmo/sageseqr/workflows/R-CMD-check/badge.svg" alt="R build status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Kelsey Montgomery, Jake Gockley.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
