<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ sageseqr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sageseqr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/customize-config.html">Write your own config file</a>
    </li>
    <li>
      <a href="../articles/run-the-plan.html">Run the targets plan</a>
    </li>
    <li>
      <a href="../articles/sagseqr-alpha-steps.html">Run the drake plan in version 0.0.0.9-alpha</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Sage-Bionetworks/sageseqr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="association_statistic_for_factors.html">association_statistic_for_factors()</a></code> </p>
        </td>
        <td><p>Compute the association statistic for factor variables</p></td>
      </tr><tr><td>
          <p><code><a href="association_statistics_for_both.html">association_statistics_for_both()</a></code> </p>
        </td>
        <td><p>Find intraclass correlation (ICC) between factor and continuous covariates</p></td>
      </tr><tr><td>
          <p><code><a href="biomart_obj.html">biomart_obj()</a></code> </p>
        </td>
        <td><p>Get available Ensembl dataset</p></td>
      </tr><tr><td>
          <p><code><a href="biomart_stats.html">biomart_stats()</a></code> </p>
        </td>
        <td><p>Get Gene Length and GC: biomart</p></td>
      </tr><tr><td>
          <p><code><a href="boxplot_vars.html">boxplot_vars()</a></code> </p>
        </td>
        <td><p>Explore metadata by variable.</p></td>
      </tr><tr><td>
          <p><code><a href="build_formula.html">build_formula()</a></code> </p>
        </td>
        <td><p>Formula Builder</p></td>
      </tr><tr><td>
          <p><code><a href="check_mismatch.html">check_mismatch()</a></code> </p>
        </td>
        <td><p>Check for consistent identifiers</p></td>
      </tr><tr><td>
          <p><code><a href="clean_covariates.html">clean_covariates()</a></code> </p>
        </td>
        <td><p>Create covariate matrix from tidy metadata data frame.</p></td>
      </tr><tr><td>
          <p><code><a href="coerce_continuous.html">coerce_continuous()</a></code> </p>
        </td>
        <td><p>Coerce objects to type numeric.</p></td>
      </tr><tr><td>
          <p><code><a href="coerce_factors.html">coerce_factors()</a></code> </p>
        </td>
        <td><p>Coerce objects to type factors.</p></td>
      </tr><tr><td>
          <p><code><a href="collapse_duplicate_hgnc_symbol.html">collapse_duplicate_hgnc_symbol()</a></code> </p>
        </td>
        <td><p>Duplicate HGNC</p></td>
      </tr><tr><td>
          <p><code><a href="complete_path.html">complete_path()</a></code> </p>
        </td>
        <td><p>Create complete URL to Synapse entity</p></td>
      </tr><tr><td>
          <p><code><a href="compute_residuals.html">compute_residuals()</a></code> </p>
        </td>
        <td><p>Compute residualized counts matrix</p></td>
      </tr><tr><td>
          <p><code><a href="conditional_plot_sexcheck.html">conditional_plot_sexcheck()</a></code> </p>
        </td>
        <td><p>Conditionally wrap plot_sexcheck for targets</p></td>
      </tr><tr><td>
          <p><code><a href="convert_geneids.html">convert_geneids()</a></code> </p>
        </td>
        <td><p>Get gene Ids</p></td>
      </tr><tr><td>
          <p><code><a href="correlate_and_plot.html">correlate_and_plot()</a></code> </p>
        </td>
        <td><p>Calculate correlation between covariates and principal components (PCs)</p></td>
      </tr><tr><td>
          <p><code><a href="cqn.html">cqn()</a></code> </p>
        </td>
        <td><p>Conditional Quantile Normalization (CQN)</p></td>
      </tr><tr><td>
          <p><code><a href="differential_expression.html">differential_expression()</a></code> </p>
        </td>
        <td><p>Differential Expression with Dream</p></td>
      </tr><tr><td>
          <p><code><a href="dropped_genes.html">dropped_genes()</a></code> </p>
        </td>
        <td><p>Dropped Genes Finder</p></td>
      </tr><tr><td>
          <p><code><a href="filter_genes.html">filter_genes()</a></code> </p>
        </td>
        <td><p>Filter genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_association_statistics.html">get_association_statistics()</a></code> </p>
        </td>
        <td><p>Visualize relationship between variables</p></td>
      </tr><tr><td>
          <p><code><a href="get_biomart.html">get_biomart()</a></code> </p>
        </td>
        <td><p>Get Ensembl biomaRt object</p></td>
      </tr><tr><td>
          <p><code><a href="get_data.html">get_data()</a></code> </p>
        </td>
        <td><p>Detect file type and download data</p></td>
      </tr><tr><td>
          <p><code><a href="gtf_stats.html">gtf_stats()</a></code> </p>
        </td>
        <td><p>Get Gene Length and GC: GTF/Fasta</p></td>
      </tr><tr><td>
          <p><code><a href="identify_outliers.html">identify_outliers()</a></code> </p>
        </td>
        <td><p>Explore samples that are outliers</p></td>
      </tr><tr><td>
          <p><code><a href="parse_counts.html">parse_counts()</a></code> </p>
        </td>
        <td><p>Subset counts data frame</p></td>
      </tr><tr><td>
          <p><code><a href="plot_coexpression.html">plot_coexpression()</a></code> </p>
        </td>
        <td><p>Co-expression histogram generation</p></td>
      </tr><tr><td>
          <p><code><a href="plot_pcs_with_covariates.html">plot_pcs_with_covariates()</a></code> </p>
        </td>
        <td><p>Explore relationship between principal components (PCs) and covariates</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sexcheck.html">plot_sexcheck()</a></code> </p>
        </td>
        <td><p>Explore metadata by gene expression on the sex chromosomes.</p></td>
      </tr><tr><td>
          <p><code><a href="plot_sexcheck_pca.html">plot_sexcheck_pca()</a></code> </p>
        </td>
        <td><p>Explore metadata by gene expression on the sex chromosomes (PCA)</p></td>
      </tr><tr><td>
          <p><code><a href="plot_volcano.html">plot_volcano()</a></code> </p>
        </td>
        <td><p>Volcano plot differential expression results'</p></td>
      </tr><tr><td>
          <p><code><a href="prepare_results.html">prepare_results()</a></code> </p>
        </td>
        <td><p>Prepare output</p></td>
      </tr><tr><td>
          <p><code><a href="provenance_helper.html">provenance_helper()</a></code> </p>
        </td>
        <td><p>Provenance helper</p></td>
      </tr><tr><td>
          <p><code><a href="run_pca.html">run_pca()</a></code> </p>
        </td>
        <td><p>Principal Component Analysis (PCA) of samples</p></td>
      </tr><tr><td>
          <p><code><a href="run_pca_and_plot_correlations.html">run_pca_and_plot_correlations()</a></code> </p>
        </td>
        <td><p>Compute principal component analysis (PCA) and plot correlations</p></td>
      </tr><tr><td>
          <p><code><a href="simple_filter.html">simple_filter()</a></code> </p>
        </td>
        <td><p>Filter genes with low expression</p></td>
      </tr><tr><td>
          <p><code><a href="start_de.html">start_de()</a></code> </p>
        </td>
        <td><p>Initialize differential expression analysis workflow</p></td>
      </tr><tr><td>
          <p><code><a href="stepwise_regression.html">stepwise_regression()</a></code> </p>
        </td>
        <td><p>Stepwise Regression</p></td>
      </tr><tr><td>
          <p><code><a href="store_results.html">store_results()</a></code> </p>
        </td>
        <td><p>Store output to Synapse</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_biotypes.html">summarize_biotypes()</a></code> </p>
        </td>
        <td><p>Summarize Biotypes</p></td>
      </tr><tr><td>
          <p><code><a href="threshold_fdr.html">threshold_fdr()</a></code> </p>
        </td>
        <td><p>Apply false discovery rate (FDR) threshold to correlation matrix.</p></td>
      </tr><tr><td>
          <p><code><a href="wrap_de.html">wrap_de()</a></code> </p>
        </td>
        <td><p>Wrapper for Differential Expression Analysis</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kelsey Montgomery, Jake Gockley, Daisy Han.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

