<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Wrapper for Differential Expression Analysis — wrap_de • sageseqr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Wrapper for Differential Expression Analysis — wrap_de"><meta property="og:description" content='This function will pass multiple conditions to test to "sagseqr::differential_expression()".'><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sageseqr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/customize-config.html">Write your own config file</a>
    </li>
    <li>
      <a href="../articles/run-the-plan.html">Run the targets plan</a>
    </li>
    <li>
      <a href="../articles/sagseqr-alpha-steps.html">Run the drake plan in version 0.0.0.9-alpha</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Sage-Bionetworks/sageseqr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Wrapper for Differential Expression Analysis</h1>
    <small class="dont-index">Source: <a href="https://github.com/Sage-Bionetworks/sageseqr/blob/HEAD/R/functions.R" class="external-link"><code>R/functions.R</code></a></small>
    <div class="hidden name"><code>wrap_de.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function will pass multiple conditions to test to <code>"sagseqr::differential_expression()"</code>.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">wrap_de</span><span class="op">(</span>
  <span class="va">conditions</span>,
  <span class="va">filtered_counts</span>,
  <span class="va">cqn_counts</span>,
  <span class="va">md</span>,
  <span class="va">dropped</span>,
  <span class="va">biomart_results</span>,
  <span class="va">p_value_threshold</span>,
  <span class="va">fold_change_threshold</span>,
  model_variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">md</span><span class="op">)</span>,
  random_effect <span class="op">=</span> <span class="cn">NULL</span>,
  exclude_variables <span class="op">=</span> <span class="cn">NULL</span>,
  cores <span class="op">=</span> <span class="cn">NULL</span>,
  cont <span class="op">=</span> <span class="cn">NULL</span>,
  is_num <span class="op">=</span> <span class="cn">NULL</span>,
  num_var <span class="op">=</span> <span class="cn">NULL</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>conditions</dt>
<dd><p>A named list of conditions to test as `primary`
in <code>"sagseqr::differential_expression()"</code>.</p></dd>
<dt>filtered_counts</dt>
<dd><p>A counts data frame with genes removed that have low expression.</p></dd>
<dt>cqn_counts</dt>
<dd><p>A counts data frame normalized by CQN.</p></dd>
<dt>md</dt>
<dd><p>A data frame with sample identifiers in a column and relevant experimental covariates.</p></dd>
<dt>dropped</dt>
<dd><p>a vector of gene names to drop from filtered counts, as they
were not cqn normalized</p></dd>
<dt>biomart_results</dt>
<dd><p>Output of <code>"sageseqr::get_biomart()"</code>. Gene Ids are
stored as rownames.</p></dd>
<dt>p_value_threshold</dt>
<dd><p>Numeric. P-values are adjusted by Benjamini and
Hochberg (BH) false discovery rate (FDR). Significant genes are those with an
adjusted p-value greater than this threshold.</p></dd>
<dt>fold_change_threshold</dt>
<dd><p>Numeric. Significant genes are those with a
fold-change greater than this threshold.</p></dd>
<dt>model_variables</dt>
<dd><p>Optional. Vector of variables to include in the linear (mixed) model.
If not supplied, the model will include all variables in <code>md</code>.</p></dd>
<dt>random_effect</dt>
<dd><p>A vector of variables to consider as random effects instead
of fixed effects.</p></dd>
<dt>exclude_variables</dt>
<dd><p>Vector of variables to exclude from testing.</p></dd>
<dt>cores</dt>
<dd><p>An integer of cores to specify in the parallel backend (eg. 4).</p></dd>
<dt>cont</dt>
<dd><p>Optional. A list specifying contrasts of the primary variable(s)
to consider for differential sequencing results if using factor(s) as your
primary variable. If not specified all combinations will be tested. If specified
this will speed up the pipeline. Specify the contrast with the factor values
involved in the contrast separated by a hyphen. (eg for diagnosis, <code>"primary_variable = c(\"AD-CT\")"</code>
where AD is the value in diagnosis column for cases and CT is the value for controls.
For multi-level contrasts, eg. <code>"primary_variable = c(\"diagnosis\", \"Sex\")"</code> would have contrasts
specified as; <code>cont= c(\"ZZ_F-CT_F\", \"ZZ_M-CT_M\")</code> to look at cases vs controls in
females and cases vs controls in males independently. While the order before or
after the hyphen doesn't matter, the order of values  before/after the underscore does matter.
The value order must be the same as the <code>"primary:} specification.
eg. \code{"primary_variable = c(\"diagnosis\",\"sex\")</code> must be <code>\"CT_M-ZZ_M\"</code> while
<code>"primary_variable = c(\"sex\",\"diagnosis\")"</code> must be <code>\"M_CT-M_ZZ\"</code> (default= NULL)</p></dd>
<dt>is_num</dt>
<dd><p>Is there a numerical covariate to use as an interaction with the primary variable(s). default= NULL</p></dd>
<dt>num_var</dt>
<dd><p>A numerical metadata column to use in an inaction with the primary variable(s). default= NULL</p></dd>
</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kelsey Montgomery, Jake Gockley, Daisy Han.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

