---
title: "sageseqr-report"
date: "`r date()`"
output: 
  html_document:
    css: style.css
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      fig.align = 'left',
                      fig.width = 5, 
                      fig.height = 5)
library(sageseqr)
```
# Data


# Explore Metadata

```{r sample_summary, message = FALSE}
drake::loadd(clean_md)
var <- config::get("x_var")
summary <- dplyr::group_by_at(clean_md, var)
summary <- dplyr::summarise(summary, count = dplyr::n())
knitr::kable(summary)
```

Visualize the distribution of data.
```{r boxplot}
drake::readd(boxplots)
```

```{r boxplot_sex_chromosomes, results = "asis"}
if (!is.null(config::get("sex check"))) {
  cat("Visualize gene expression (log-transformed counts) of sex chromosomes using XIST 
      as a X-chromosome marker and UTY as a Y-chromosome marker.")
  drake::readd(sex_plot)$plot
}
```

Visualize the relationships between covariates.
```{r heatmap_covariates}
correlation_input <- drake::readd(correlation_plot)$plot
col2 <- grDevices::colorRampPalette(rev(c("#67001F", "#B2182B", "#D6604D", "#F4A582",
                                 "#FDDBC7", "#FFFFFF", "#D1E5F0", "#92C5DE",
                                 "#4393C3", "#2166AC", "#053061")))
corrplot::corrplot(correlation_input, col = col2(200), tl.col = "#000000")
```
# Filter Genes
```{r filter_genes}
`r dim(PROCESSED_COUNTS$filteredExprMatrix$genes)[1]` genes are used for the analysis. 

Following is the distribution of biotypes
`r kable(pct.pc)`
```

# Identify Outliers
```{r plot_outliers}
```

