<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Write your own config file â€¢ sageseqr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Write your own config file">
<meta property="og:description" content="sageseqr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sageseqr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/customize-config.html">Write your own config file</a>
    </li>
    <li>
      <a href="../articles/run-the-plan.html">Run the targets plan</a>
    </li>
    <li>
      <a href="../articles/sagseqr-alpha-steps.html">Run the drake plan in version 0.0.0.9-alpha</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Sage-Bionetworks/sageseqr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="customize-config_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Write your own config file</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Sage-Bionetworks/sageseqr/blob/HEAD/vignettes/customize-config.Rmd" class="external-link"><code>vignettes/customize-config.Rmd</code></a></small>
      <div class="hidden name"><code>customize-config.Rmd</code></div>

    </div>

    
    
<p>A <a href="https://github.com/kelshmo/sageseqr/blob/master/config.yml" class="external-link">configuration file</a> defines the inputs for the RNA-seq workflow. Update the default parameters to include the Synapse ID where your data is stored and to the factor and continuous variables you want to test in the covariate model selection. The full list of configurable options are:</p>
<pre><code>counts:
  synID:          Required. Synapse ID to counts data frame with identifiers to 
                  the metadata as column names and gene ids in a column.
  version:        Optional. Include Synapse file version number (e.g. 3).
  gene id:        Required. Column name that corresponds to the gene ids (e.g. feature).
metadata:
  synID:          Required. Synapse ID to cleaned metadata file with sample 
                  identifiers in a column and variables of interest as column names. 
  version:        Optional. Include Synapse file version number (e.g. 3).
  sample id:      Required. Column name that corresponds to the sample ids (e.g. donorid).
biomart:
  synID:          Optional. If left blank, Ensembl will be queried with the gene 
                  ids provided in the counts. Otherwise, you may provide the
                  Synapse ID to gene metadata from Ensembl. This must include gene
                  length and GC content in order to implement Conditional Quantile
                  Normalization. 
  version:        Optional. Include Synapse file version number (e.g. 3).
  filters:        Required. Column name that corresponds to the gene ids (e.g. 
                  ensembl_gene_id).
  host:           Optional. A character vector specifying the BioMart database release
                  version. This specification is highly recommended for a reproducible           
                  workflow. Defaults to ensembl.org.
  organism:       Required. A character vector of the organism name. This argument 
                  takes partial strings. For example,"hsa" will match "hsapiens_gene_ensembl".
  exon only:      Optional. Set to TRUE if you want gene lengths and GC-content
                  to be calculated only for exons of gene features. Recomended 
                  depending on you experimental design paradigm. Default is FALSE,
                  which considers the entire transcript start to stop (ie. includes
                  intronic regions).
  custom build:   Optional. If you want to bulid the biomart object from a user specified
                  or custom GTF and genome FASTA file specify this value as TRUE.
                  Default is FALSE. This would be reccomended for users analyzing
                  data from a model system with a trans-gene inserted into the 
                  genome.
  gtfID:          Required IF custom build is set to TRUE. Synapse ID to the user
                  specified GTF file to build the biomart object from.
  gtfVersion:     Optional.Include Synapse file version number (e.g. 3).
  fastaID:        Required IF custom build is set to TRUE. Synapse ID to the user
                  specified genome FASTA file to build the biomart object from.
  fastaVersion:   Optional. Include Synapse file version number (e.g. 3).
factors:          Required. List of factor variables in brackets. Variables must be 
                  present in the metadata as column names (e.g. [ "donorid", "source"]).
continuous:       Required. List of continuous variables in brackets. Variables must 
                  be present in the metadata as column names (e.g. [ "rin", "rin2"]).
x_var:            Required. This is your predictor or primary variable of interest. 
                  Additionally, a boxplot will visualize the distribution of 
                  continuous variables using the x_var as a dimension.
conditions:       Optional. Filtering low-expression genes is a common practice to improve
                  sensitivity in detection of differentially expressed genes. Low
                  count genes that have less than a user-defined threshold of counts 
                  per million (CPM) in a user-defined percentage of samples per the
                  conditions provided here will be removed. (e.g. ["diagnosis", "sex"]). 
cpm threshold:    Optional. The minium allowable CPM to keep a gene in the analysis.
percent threshold:Optional. The percentage of samples that should contain the minimum number
                  of CPM. If a condition is passed, the percentage will apply to 
                  the samples in that sub-population. 
sex check:        Optional. The exact variable name that corresponds to reported gender or sex
                  to check the distribution of x and y marker expression across
                  samples.
dimensions:       Required. Specify the PCA dimensions by variable name. 
  color:
  shape:
  size:
skip model:       Optional. If TRUE, the exploratory data report is run. Model selection
                  is not computed.
force model with: Optional. Force differential expression with this user defined model
                  instead of the output of stepwise regression.
cores:            Optional. Specify an integer of cores to use with a BiocParallel 
                  parallel backend.  Null value results in the number of available
                  cores minus one being used. Parallel backend ccurrently only supports
                  BiocParallel::SnowParam(). BatchtoolsParam, MulticoreParam, 
                  BiocParallelDoparParam, and SerialParam are not currently supported.
de FC:            The fold-change (FC) of significant differentially expressed 
                  (de) genes must exceed this value. This value will be transformed
                  into log2 FC.
de p-value threshold: The adjusted p-value of significant differentially expressed 
                  (de) genes must exceed this value.
de contrasts:     Required. Variable in the metadata to define comparisons between 
                  groups.
  &lt;any named list:&gt; If there are multiple comparisons, set them up as nested lists.
visualization gene list: Label this list of genes in the volcano plot.
report:           Required. The name of your project. This will become the name of your 
                  output html file.
store output:     Required. Folder Synapse Id to store output on Synapse.</code></pre>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kelsey Montgomery, Jake Gockley, Daisy Han.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
